junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/DynamicTestStreamClosedTests.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.jupiter.engine.descriptor;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.lang.reflect.Method;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestFactory;
+import org.junit.jupiter.api.extension.TestExtensionContext;
+import org.junit.jupiter.engine.execution.JupiterEngineExecutionContext;
+import org.junit.jupiter.engine.execution.ThrowableCollector;
+import org.junit.platform.engine.UniqueId;
+
+public class DynamicTestStreamClosedTests {
+	private JupiterEngineExecutionContext context;
+	private Method testMethod;
+	private TestExtensionContext testExtensionContext;
+	private boolean isClosed;
+
+	@BeforeEach
+	public void before() {
+		testExtensionContext = mock(TestExtensionContext.class);
+		isClosed = false;
+
+		context = new JupiterEngineExecutionContext(null, null).extend().withThrowableCollector(
+			new ThrowableCollector()).withExtensionContext(testExtensionContext).build();
+	}
+
+	@Test
+	public void streamsFromTestFactoriesShouldBeClosed() throws NoSuchMethodException {
+		testMethod = DynamicCloseHookedStreamTest.class.getMethod("customStream");
+		when(testExtensionContext.getTestMethod()).thenReturn(Optional.of(testMethod));
+
+		Stream<DynamicTest> stream = Stream.<DynamicTest> of().onClose(() -> isClosed = true);
```

Comment:
I am fine with having it changed to `Stream.{<DynamicTest> }empty()` for explicitness's sake. :)

And if using 2 or more variables would allow the `<DynamicTest>` type arg to be removed, then yes I would personally prefer that.

It's not clear to be if adding a no-op dynamic test in the `of()` would be any more readable than using 2+ vars (I'm writing this with my phone, so I don't have access to IntelliJ IDEA, ATM), but if you think that it is more readable, then I'm happy to trust your judgement. :)